trace_decoder/src/imp.rs

```diff
@@ -0,0 +1,641 @@
+            it
+        });
+        ensure!(
+            storage.root() == acct.storage_root,
+            "inconsistent initial storage for hashed address {haddr:x}"
+        )
+    }
+
+    // These are the per-block tries.
+    let mut transaction_trie = TransactionTrie::new();
+    let mut receipt_trie = ReceiptTrie::new();
+
+    let mut out = vec![];
+
+    let mut txn_ix = 0; // incremented for non-dummy transactions
+    let mut loop_ix = 0; // always incremented
+    let loop_len = batches.iter().flatten().count();
+    for batch in batches {
+        let batch_first_txn_ix = txn_ix; // GOTCHA: if there are no transactions in this batch
+        let mut batch_gas_used = 0;
+        let mut batch_byte_code = vec![];
+        let mut batch_contract_code = BTreeSet::from([vec![]]); // always include empty code
+
+        let mut before = IntraBlockTries {
+            state: state_trie.clone(),
+            transaction: transaction_trie.clone(),
+            receipt: receipt_trie.clone(),
+            storage: storage_tries.clone(),
+        };
+
+        // We want to trim the TrieInputs above,
+        // but won't know the bounds until after the loop below,
+        // so store that information here.
+        let mut storage_masks = BTreeMap::<_, BTreeSet<TrieKey>>::new();
+        let mut state_mask = BTreeSet::new();
+
+        if txn_ix == 0 {
+            cancun_hook(
+                block_timestamp,
+                &mut storage_tries,
+                &mut storage_masks,
+                parent_beacon_block_root,
+                &mut state_mask,
+                &mut state_trie,
+            )?;
+        }
+
+        for txn in batch {
+            let do_increment_txn_ix = txn.is_some();
+            let TxnInfo {
+                traces,
+                meta:
+                    TxnMeta {
+                        byte_code,
+                        new_receipt_trie_node_byte,
+                        gas_used: txn_gas_used,
+                    },
+            } = txn.unwrap_or_default();
+
+            if let Ok(nonempty) = nunny::Vec::new(byte_code) {
+                batch_byte_code.push(nonempty.clone());
+                transaction_trie.insert(txn_ix, nonempty.into())?;
+                receipt_trie.insert(
+                    txn_ix,
+                    map_receipt_bytes(new_receipt_trie_node_byte.clone())?,
+                )?;
+            }
+
+            batch_gas_used += txn_gas_used;
+
+            for (
+                addr,
+                just_access,
+                TxnTrace {
+                    balance,
+                    nonce,
+                    storage_read,
+                    storage_written,
+                    code_usage,
+                    self_destructed,
+                },
+            ) in traces.into_iter().map(|(addr, trc)| {
+                (
+                    addr,
+                    // This is our fork of Erigon's way of letting us know
```

Comment:
We should remove the mention of Erigon here. This is also how the native tracer handles these cases, see `process_transactions` in the `native` module